File: cpp/alerts/url_seed.cpp
#include "./url_seed.h"
#include "cpp/error.h"

namespace ltrs {

rust::Str url_seed_alert_get_server_url(lt::url_seed_alert* alert) {
    return alert->server_url();
}

rust::Str url_seed_alert_get_error_message(lt::url_seed_alert* alert) {
    return alert->error_message();
}

Error url_seed_alert_get_error(lt::url_seed_alert* alert) {
    return error_code_to_error(alert->error);
}

} // namespace ltrs

File: src/ffi/alerts/url_seed.rs
#[cxx::bridge(namespace = "ltrs")]
pub(crate) mod ffi {
    #[namespace = "libtorrent"]
    extern "C++" {
        type url_seed_alert = crate::ffi::ffi::url_seed_alert;
    }

    unsafe extern "C++" {
        include!("cpp/alerts/url_seed.h");
        type Error = crate::ffi::error::ffi::Error;
        unsafe fn url_seed_alert_get_server_url<'a>(alert: *mut url_seed_alert) -> &'a str;
        unsafe fn url_seed_alert_get_error_message<'a>(alert: *mut url_seed_alert) -> &'a str;
        unsafe fn url_seed_alert_get_error(alert: *mut url_seed_alert) -> Error;
    }
}
